packages/@aws-cdk/assets/lib/asset.ts

```diff
@@ -84,15 +89,15 @@ export class Asset extends cdk.Construct {
             description: `S3 bucket for asset "${this.path}"`,
         });
 
-        const keyParam = new cdk.Parameter(this, 'S3ObjectKey', {
+        const keyParam = new cdk.Parameter(this, 'S3VersionKey', {
             type: 'String',
-            description: `S3 object for asset "${this.path}"`
+            description: `S3 key for asset version "${this.path}"`
         });
 
         this.s3BucketName = bucketParam.value;
-        this.s3ObjectKey = keyParam.value;
-
-        // grant the lambda's role read permissions on the code s3 object
+        this.s3Prefix = new cdk.FnSelect(0, new cdk.FnSplit(cxapi.ASSET_PREFIX_SEPARATOR, keyParam.value));
+        const s3Filename = new cdk.FnSelect(1, new cdk.FnSplit(cxapi.ASSET_PREFIX_SEPARATOR, keyParam.value));
```

Comment:
You realize (however unlikely that is), that `||` might actually be part of the object key?

Reply:
That is going to be true for any separator we pick. So yeah, I'm balancing aesthetics vs likelihood

Reply:
That is going to be true for any separator we pick. So yeah, I'm balancing aesthetics vs likelihood

Reply:
You do realize that git commit hashes may clash? :p